src/common/Cryptography/CryptoHash.h

```diff
@@ -0,0 +1,120 @@
+
+        template <GenericHashImpl::HashCreator HashCreator, size_t DigestLength>
+        class GenericHash
+        {
+            public:
+                static constexpr size_t DIGEST_LENGTH = DigestLength;
+                using Digest = std::array<uint8, DIGEST_LENGTH>;
+
+                static Digest GetDigestOf(uint8 const* dta, int len)
+                {
+                    GenericHash hash;
+                    hash.UpdateData(dta, len);
+                    hash.Finalize();
+                    return hash.GetDigest();
+                }
+                template <typename... Ts>
+                static auto GetDigestOf(Ts&&... pack) -> std::enable_if_t<!(std::is_integral_v<std::decay_t<Ts>> || ...), Digest>
+                {
+                    GenericHash hash;
+                    (hash.UpdateData(std::forward<Ts>(pack)), ...);
+                    hash.Finalize();
+                    return hash.GetDigest();
+                }
+
+                GenericHash() : _ctx(GenericHashImpl::MakeCTX())
+                {
+                    int result = EVP_DigestInit_ex(_ctx, HashCreator(), nullptr);
+                    ASSERT(result == 1);
+                }
+
+                ~GenericHash()
+                {
+                    if (!_ctx)
+                        return;
+                    GenericHashImpl::DestroyCTX(_ctx);
+                    _ctx = nullptr;
+                }
+
+                void UpdateData(uint8 const* dta, size_t len)
+                {
+                    int result = EVP_DigestUpdate(_ctx, dta, len);
+                    ASSERT(result == 1);
+                }
+                void UpdateData(std::string_view str) { UpdateData(reinterpret_cast<uint8 const*>(str.data()), str.size()); }
+                void UpdateData(std::string const& str) { UpdateData(std::string_view(str)); } /* explicit overload to avoid using the container template */
+                void UpdateData(char const* str) { UpdateData(std::string_view(str)); } /* explicit overload to avoid using the container template */
+                template <typename Container> void UpdateData(Container const& c) { UpdateData(std::data(c), std::size(c)); }
+
+                void Finalize()
+                {
+                    uint32 length;
+                    int result = EVP_DigestFinal_ex(_ctx, _digest.data(), &length);
+                    ASSERT(result == 1);
+                    ASSERT(length == DIGEST_LENGTH);
+                    GenericHashImpl::DestroyCTX(_ctx);
+                    _ctx = nullptr;
+                }
+
+                Digest const& GetDigest() const { return _digest; }
+
+            private:
+                EVP_MD_CTX* _ctx;
+                Digest _digest;
+        };
+    }
+
+    namespace Crypto
+    {
+        using SHA1 = Trinity::Impl::GenericHash<EVP_sha1, Constants::SHA1_DIGEST_LENGTH_BYTES>;
+        using SHA256 = Trinity::Impl::GenericHash<EVP_sha256, Constants::SHA256_DIGEST_LENGTH_BYTES>;
```

Comment:
This is vastly irrelevant but i'd use openssl defines here since you're already using openssl types, just for consistency

Reply:
We discussed this previously. See https://github.com/TrinityCore/TrinityCore/pull/25093#discussion_r460142261 . 
That's why we ended up here.

Reply:
Fair